daemons/execd/remoted_schemas.c

```diff
@@ -0,0 +1,295 @@
+
+    if ((signo == 0) && (exitcode == 0)) {
+        const char *remote_schema_dir = pcmk__env_option(PCMK__ENV_REMOTE_SCHEMA_DIR);
+
+        if (remote_schema_dir == NULL) {
+            remote_schema_dir = PCMK__REMOTE_SCHEMA_DIR;
+        }
+
+        /* Don't just crm_schema_init here because that will load the base
+         * schemas again too.  Instead just load the things we fetched.
+         */
+        pcmk__load_schemas_from_dir(remote_schema_dir);
+
+    } else {
+        if (signo == 0) {
+            crm_err("%s: process %d exited %d", errmsg, (int) pid, exitcode);
+
+        } else {
+            crm_err("%s: process %d terminated with signal %d (%s)%s",
+                    errmsg, (int) pid, signo, strsignal(signo),
+                    (core? " and dumped core" : ""));
+        }
+
+        /* Clean up any incomplete schema data we might have been downloading when
+         * the process timed out or crashed.  We don't need to do any extra cleanup
+         * because we never loaded the extra schemas, and we don't need to call
+         * crm_schema_init because that was called in remoted_request_cib_schema_files
+         * before this function.
+         */
+        clean_up_extra_schema_files();
+    }
+}
+
+void
+remoted_request_cib_schema_files(void)
+{
+    pid_t pid;
+    gboolean killed = FALSE;
+    int rc;
+
+    /* If a previous schema-fetch process is still running when we're called
+     * again, it's hung.  Attempt to kill it before cleaning up the extra
+     * directory.
+     */
+    if (schema_fetch_pid != 0) {
+        mainloop_child_kill(schema_fetch_pid);
+
+        if (killed == FALSE) {
+            crm_warn("Unable to kill pre-existing schema-fetch process");
+            return;
+        }
+
+        schema_fetch_pid = 0;
+    }
+
+    /* Clean up any extra schema files we downloaded from a previous cluster
+     * connection.  After the files are gone, we need to wipe them from
+     * known_schemas, but there's no opposite operation for add_schema().
+     *
+     * Instead, unload all the schemas.  This means we'll also forget about all
+     * the installed schemas as well, which means that xml_latest_schema() will
+     * fail.  So we need to load the base schemas right now.
+     */
+    clean_up_extra_schema_files();
+    crm_schema_cleanup();
+    crm_schema_init();
+
+    pid = fork();
+
+    switch (pid) {
+        case -1: {
+            rc = errno;
+            crm_warn("Could not spawn process to get schema files: %s", pcmk_rc_str(rc));
+            break;
+        }
+
+        case 0:
+            /* child */
+            get_schema_files();
+            break;
+
+        default:
+            /* parent */
+            schema_fetch_pid = pid;
+            mainloop_child_add_with_flags(pid, 5 * 60 * 1000, "schema-fetch", NULL,
+                                          mainloop_leave_pid_group,
+                                          get_schema_files_complete);
```

Comment:
Probably worth logging that we have requested any new schema files, and then log when it completes. That way if anyone has any problems in the time between, they can figure out why (or if there were no new schemas, they can rule that out).